app/services/the_ingestor.rb

```diff
@@ -9,33 +9,43 @@ def ingest
 
     @document.chunking!
     chonker.each_with_index do |chunk_record, index|
-      if index.zero?
-        # Temporary hack since parser is still a brute force chunker right now.
-        @document.chunked!
-        @document.embedding!
-      end
-
-      chunk = Chunk.new(document: @document, content: chunk_record.content)
-      chunk.save!
-
-      Rails.logger.info("Embedding chunk #{chunk.id} (#{index})")
-      embedding = embedder.embed(chunk_record.embedding_content)
-
-      ids = chromadb.add_documents(@collection_id, [chunk.content], [embedding])
-      chunk.update!(vector_id: ids.first)
+      prepare_and_embed_chunk(chunk_record, index)
     end
     Rails.logger.info("Got chunks from #{@document.filename}.")
 
     @document.embedded!
   rescue StandardError => e
-    Rails.logger.error("Error ingesting document #{@document.id}\n#{e.class.name}: #{e.message}#{e.backtrace.join("\n")}")
+    Rails.logger.error("Error ingesting document #{@document.id}\n#{exception_summary(e)}")
     @document.error!
 
     raise e
   end
 
   private
 
+  def exception_summary(err)
+    "#{err.class.name}: #{err.message}#{err.backtrace.join("\n")}"
+  end
+
+  def document_embedding!
+    # Temporary hack since parser is still a brute force chunker right now.
+    @document.chunked!
+    @document.embedding!
+  end
+
+  def prepare_and_embed_chunk(chunk_record, index)
+    document_embedding! if index.zero?
```

Comment:
Could you do this once, before the outer `each_with_index`?

Reply:
I call `#chunking!` outside the loop and then this inside to track `#chunked!` because it's possible that `chonker.each_with_index` could fail after "chunking" but before "chunked."

Pedantic? Yes.

Reply:
Actually, I have refactor for this, coming shortly.